File: Day #1673/solution.js
// Day 1673: Min wheel rotations from "000" to target avoiding dead ends.
// BFS over <=1000 states, each with 6 neighbors. Time O(1000), Space O(1000).
function openLock(target, deadends) {
  const dead = new Set(deadends);
  const start = "000";
  if (dead.has(start) || dead.has(target)) return null;
  if (start === target) return 0;
  const dist = new Map([[start, 0]]);
  const q = [start];
  let head = 0;
  while (head < q.length) {
    const cur = q[head++];
    for (let i = 0; i < 3; i++) {
      for (const d of [-1, 1]) {
        const digit = (Number(cur[i]) + d + 10) % 10;
        const nxt = cur.slice(0, i) + digit + cur.slice(i + 1);
        if (dead.has(nxt) || dist.has(nxt)) continue;
        dist.set(nxt, dist.get(cur) + 1);
        if (nxt === target) return dist.get(nxt);
        q.push(nxt);
      }
    }
  }
  return null;
}

console.log(openLock("345", ["333"])); // 12

File: Day #1673/solution.rs
// Day 1673: Min wheel rotations from "000" to target avoiding dead ends.
// BFS over <=1000 states, each with 6 neighbors. Time O(1000), Space O(1000).
use std::collections::{HashMap, HashSet, VecDeque};

fn open_lock(target: &str, deadends: &[&str]) -> Option<i32> {
    let dead: HashSet<String> = deadends.iter().map(|s| s.to_string()).collect();
    let start = "000".to_string();
    if dead.contains(&start) || dead.contains(target) {
        return None;
    }
    if start == target {
        return Some(0);
    }
    let mut dist: HashMap<String, i32> = HashMap::new();
    dist.insert(start.clone(), 0);
    let mut q: VecDeque<String> = VecDeque::new();
    q.push_back(start);
    while let Some(cur) = q.pop_front() {
        let cd = dist[&cur];
        let bytes = cur.as_bytes();
        for i in 0..3 {
            for d in [-1i32, 1] {
                let digit = (bytes[i] as i32 - '0' as i32 + d + 10) % 10;
                let mut nb = bytes.to_vec();
                nb[i] = b'0' + digit as u8;
                let nxt = String::from_utf8(nb).unwrap();
                if dead.contains(&nxt) || dist.contains_key(&nxt) {
                    continue;
                }
                dist.insert(nxt.clone(), cd + 1);
                if nxt == target {
                    return Some(cd + 1);
                }
                q.push_back(nxt);
            }
        }
    }
    None
}

fn main() {
    match open_lock("345", &["333"]) {
        Some(r) => println!("{}", r), // 12
        None => println!("None"),
    }
}
